src/util/circular_buffer.c

```diff
@@ -0,0 +1,259 @@
+  RedisModule_Assert(cb != NULL);
+
+  return atomic_load(&cb->item_count);
+}
+
+// Returns buffer capacity.
+uint64_t CircularBuffer_Cap(CircularBuffer cb) {
+  RedisModule_Assert(cb != NULL);
+
+  return cb->item_cap;
+}
+
+// Returns the size of each item in the buffer.
+uint CircularBuffer_ItemSize(const CircularBuffer cb) {
+  return cb->item_size;
+}
+
+// Returns true if buffer is empty. Thread-safe.
+inline bool CircularBuffer_Empty(const CircularBuffer cb) {
+  RedisModule_Assert(cb != NULL);
+
+  uint64_t item_count = atomic_load(&cb->item_count);
+  return item_count == 0;
+}
+
+// Returns true if buffer is full. Thread-safe.
+inline bool CircularBuffer_Full(const CircularBuffer cb) {
+  RedisModule_Assert(cb != NULL);
+
+  uint64_t item_count = atomic_load(&cb->item_count);
+  return item_count == cb->item_cap;
+}
+
+// Adds an item to buffer.
+// Returns 1 on success, 0 otherwise
+// This function is thread-safe and lock-free
+int CircularBuffer_Add(CircularBuffer cb, void *item) {
+  RedisModule_Assert(cb   != NULL);
+  RedisModule_Assert(item != NULL);
+
+  // atomic update buffer item count
+  // do not add item if buffer is full
+  uint64_t item_count = atomic_fetch_add(&cb->item_count, 1);
+  if (unlikely(item_count >= cb->item_cap)) {
+    atomic_fetch_sub(&cb->item_count, 1);
+    return 0;
+  }
+
+  // determine current and next write position
+  uint64_t offset = atomic_fetch_add(&cb->write, cb->item_size);
+
+  // check for buffer overflow
+  if (unlikely(cb->data + offset >= cb->end_marker)) {
+    // write need to circle back
+    // [., ., ., ., ., ., A, B, C]
+    //                           ^  ^
+    //                           W0 W1
+    uint64_t overflow = offset + cb->item_size;
+
+    // adjust offset
+    // [., ., ., ., ., ., A, B, C]
+    //  ^  ^
+    //  W0 W1
+    offset -= cb->item_size * cb->item_cap;
+
+    // update write position
+    // multiple threads "competing" to update write position
+    // we ensure that the thread with the largest offset will succeed
+    // for the above example, W1 will succeed
+    //
+    // [., ., ., ., ., ., A, B, C]
+    //        ^
+    //        W
+    atomic_compare_exchange_strong(&cb->write, &overflow, offset + cb->item_size);
+  }
+
+  // copy item into buffer
+  memcpy(cb->data + offset, item, cb->item_size);
+
+  // report success
+  return 1;
+}
+
+// Reserve a slot within buffer.
+// Returns a pointer to a 'item size' slot within the buffer.
+// This function is thread-safe and lock-free.
+// [OUTPUT] wasFull - set to true if buffer is full
+void *CircularBuffer_Reserve(CircularBuffer cb, bool *wasFull) {
+  RedisModule_Assert(cb != NULL);
+
+  // atomic update buffer item count
+  // an item will be overwritten if buffer is full
+  uint64_t item_count = atomic_fetch_add(&cb->item_count, 1);
+  if (unlikely(item_count >= cb->item_cap)) {
+    atomic_fetch_sub(&cb->item_count, 1);
+    if (wasFull != NULL) {
+      *wasFull = true;
```

Comment:
Not setting `wasFull` to false when wasn't full?
(so the caller needs to clear `wasFull` before calling)

Reply:
Right

Reply:
So need to comment this in the header declaration?

Reply:
Added the setting of it to false as well.